keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+result in a cleaner separation at the source code level. It has no impact on
+the resulting Kubernetes API.
+
+### Scheduling + Allocation
+
+The dynamic resource scheduler plugin handles the common fields of
+NodeResourceSlice, ResourceClaimParameters and SemanticResourceHandle. For the
+semantic model fields it calls out to code that is associated with the
+corresponding model.
+
+During filtering it is decided which nodes have the necessary resources. If a
+node is found, the scheduler plugin updates the resource claim status as part
+of goroutine which handles pod binding.
+
+Like a normal DRA driver controller, the scheduler also sets a finalizer to
+ensure that users cannnot accidentally delete the allocated claim while a pod
+is about to start which depends on it. That finalizer is
+"semantic.dra.k8s.io/delete-protection".
+
+### Deallocation
+
+Deallocation is handled by kube-controller-manager when its claim controller
+observes that a claim is no longer in use *and* the claim has the special
+"semantic.dra.k8s.io/delete-protection" finalizer. This finalizer tells the
+controller that it may clear the allocation result directly instead of setting
+the `DeletionRequested` field, which is what it normally would do.
+
+Updating the claim during deallocation will be observed by kube-scheduler and
+tells it that it can use the capacity set aside for the claim
+again. kube-controller-manager itself doesn't need to support specific semantic
+models.
+
+### Immediate allocation
+
+Because there is no separate controller anymore, claims with immediate
+allocation will only get allocated once there is a pod which needs them. The
+remaining semantic difference compared to delayed allocation is that claims
+with immediate allocation remain allocated when no longer in use.
+
+### Simulation with CA
+
+**Note**: this section outlines one potential solution for Cluster Autoscaler.
+During code review it will be decided whether the actual implement is done like
+this. Other autoscalers will have to implement a similar logic if they don't
+use the in-tree dynamic resourceclaim scheduler plugin.
+
+The usual call sequence of a scheduler plugin when used in the scheduler is
+at program startup:
+- instantiate plugin
+- EventsToRegister
+
+For each new pod:
+- PreEnqueue
+
+For each pod that is ready to be scheduled, one pod at a time:
+- PreFilter, Filter, etc.
+
+
+CA works a bit differently. It identifies all pending pods,
+takes a snapshot of the current cluster state, and then simulates the effect
+of scheduling those pods with additional nodes added to the cluster. To
+determine whether a pod fits into one of these simulated nodes, it
+uses the same PreFilter and Filter plugins as the scheduler. Other extension
+points (Reserve, Bind) are not used. Plugins which modify the cluster state
+therefore need a different way of recording the result of scheduling
+a pod onto a node.
+
+This is done through a new `ClusterAutoScalerPlugin` interface defined in
```

Comment:
Hey, I'm a core maintainer for Cluster Autoscaler, working closely with @MaciekPytel.

The proposed approach for Cluster Autoscaler implementation isn't consistent with how CA integrates with the scheduler framework. The current approach is that during the simulations CA works on its own snapshot of the cluster, adding/removing nodes, and moving pods around. The snapshot implements listers for objects that scheduler plugins need (e.g. most notably `NodeInfo`), and the scheduler plugins uss these fake snapshot listers instead of talking directly to the apiserver.

It seems that we could use the same pattern for this as well:
* Allow injecting listers for `NodeResourceSlice`, `ResourceClaim`, `ResourceClaimParameters` into the scheduler framework. Or maybe we could add them to the existing framework lister for `NodeInfo` - `[]NodeResourceSlice` to `NodeInfo`, and `[]ResourceClaim`+`[]ResourceClaimParameters` to `PodInfo`. The scheduler will use real listers, CA will use the cluster snapshot fake listers.
* Allow injecting an updater for `ResourceClaim` into the scheduler framework, to handle setting `SemanticResourceHandle` for allocation. The scheduler will use a real Updater, CA will use an Updater that will just modify its snapshot.
* Extend the Cluster Autoscaler cluster snapshot to allow listing `NodeResourceSlice`, `ResourceClaim`, `ResourceClaimParameters`, and modifying `ResourceClaim`. 
* Cluster Autoscaler starts calling the `Bind` method for nodes that use DRA. This could be scoped to just the DRA's plugin `Bind`. This allows the DRA scheduler plugin to update the relevant `ResourceClaim` in CA's cluster snapshot as allocated whenever CA puts a pod onto a node.
* Cluster Autoscaler would clear the allocation in the `ResourceClaims` associated with a pod whenever it simulates moving it to a different node. CA sanitizes such a pod anyway because technically it's a new pod on the new node - sanitizing the `ResourceClaims` fits as a part of that.
* The DRA scheduler plugin doesn't recognize the difference between running in scheduler or CA - it just reads and modifies the cluster using the injected listers/updaters. In scheduler they reflect a real cluster, in CA they reflect the cluster snapshot.

@pohly WDYT about this approach? I know we're nearing a KEP deadline, I could flesh it out and update this section of the KEP myself if bandwidth is an issue. I'm also happy to take the CA side of the implementation when we get to that stage.

I'm also curious to hear your thoughts on this from the scheduler side @alculquicondor.

Reply:
@towca I'd say go ahead and propose some text to include directly in the KEP. At this stage, I think that would be the best course of action.

Reply:
Understood, I agree that we don't need to decide now. Listing possible implementations sounds good to me, thanks!